src/engine/strat_engine/engine.rs

```diff
@@ -124,6 +140,49 @@ impl Engine for StratEngine {
         Ok(uuid)
     }
 
+    /// Evaluate a device node & devicemapper::Device to see if it's a valid
+    /// stratis device.  If all the devices are present in the pool and the pool isn't already
+    /// up and running, it will get setup and the pool uuid will be returned.
+    fn block_evaluate(&mut self, dev_node: PathBuf, device: Device) -> Option<PoolUuid> {
+        let mut duplicate = false;
+
+        match is_stratis_device(&mut self.incomplete_pools,
+                                dev_node,
+                                libc::dev_t::from(device)) {
+            Ok(process_it) => {
+                let mut rc = None;
+                if let Some(pool_uuid) = process_it {
+                    if let Some(devices) = self.incomplete_pools.get(&pool_uuid) {
+                        // Check to see if we already have a pool up and running with this uuid,
+                        // if we do we are going to ignore this setup.  We can re-create this
+                        // case by creating synthetic "add" events for existing stratis devices.
+                        if !self.pools.contains_uuid(pool_uuid) {
+                            if let Ok(pool_setup) = StratPool::setup(pool_uuid, devices) {
+                                self.pools.insert(pool_setup);
+                                rc = Some(pool_uuid);
+                            }
+                        } else {
+                            // We have this pool already so we shall ignore.
+                            warn!("Pool {} is already known, ignoring!", pool_uuid);
+                            duplicate = true;
+                        }
+                    }
+
+                    // If we brought up a new pool or we found an existing pool we will remove the
+                    // information from the holding pool as we no longer need it.
+                    if rc.is_some() || duplicate {
+                        self.incomplete_pools.remove(&pool_uuid);
+                    }
+
+                    return rc;
+                }
+
+                None
+            }
+            _ => None,
```

Comment:
I think ```block_evaluate()``` should return an error if ```is_stratis_device``` returns an error. It didn't succesfully evaluate the device for being a stratis device or not. You can choose to ignore the error at a higher level with the ```_```. So structure would just be:

```
match is stratis_device(...)? {
    Some(uuid) => {
      // all the interesting stuff
    },
    None => None
}
```

And return type would be ```Err<Option<PoolUuid>>```.